crates/futures/src/nightly.rs

```diff
@@ -0,0 +1,252 @@
+            tx2.send(val).unwrap_throw();
+        });
+
+        js.then2(&cb_resolve, &cb_reject);
+
+        JsFuture {
+            resolved: rx1,
+            rejected: rx2,
+            _cb_resolve: cb_resolve,
+            _cb_reject: cb_reject,
+        }
+    }
+}
+
+impl Future for JsFuture {
+    type Output = Result<JsValue, JsValue>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // Test if either our resolved or rejected side is finished yet.
+        if let Poll::Ready(val) = self.resolved.poll_unpin(cx) {
+            return Poll::Ready(Ok(val.unwrap_throw()));
+        }
+
+        if let Poll::Ready(val) = self.rejected.poll_unpin(cx) {
+            return Poll::Ready(Err(val.unwrap_throw()));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// Converts a Rust `Future` into a JavaScript `Promise`.
+///
+/// This function will take any future in Rust and schedule it to be executed,
+/// returning a JavaScript `Promise` which can then be passed back to JavaScript
+/// to get plumbed into the rest of a system.
+///
+/// The `future` provided must adhere to `'static` because it'll be scheduled
+/// to run in the background and cannot contain any stack references. The
+/// returned `Promise` will be resolved or rejected when the future completes,
+/// depending on whether it finishes with `Ok` or `Err`.
+///
+/// # Panics
+///
+/// Note that in wasm panics are currently translated to aborts, but "abort" in
+/// this case means that a JavaScript exception is thrown. The wasm module is
+/// still usable (likely erroneously) after Rust panics.
+///
+/// If the `future` provided panics then the returned `Promise` **will not
+/// resolve**. Instead it will be a leaked promise. This is an unfortunate
+/// limitation of wasm currently that's hoped to be fixed one day!
+pub fn future_to_promise<F>(future: F) -> Promise
+where
+    F: Future<Output = Result<JsValue, JsValue>> + 'static,
+{
+    let mut future = Some(future);
+
+    Promise::new(&mut |resolve, reject| {
+        // TODO change Promise::new to be FnOnce
+        spawn_local(future.take().unwrap_throw().map(move |val| {
+            match val {
+                Ok(val) => {
+                    resolve.call1(&JsValue::undefined(), &val).unwrap_throw();
+                },
+                Err(val) => {
+                    reject.call1(&JsValue::undefined(), &val).unwrap_throw();
+                },
+            }
+        }));
+    })
+}
+
+/// Runs a Rust `Future` on a local task queue.
+///
+/// The `future` provided must adhere to `'static` because it'll be scheduled
+/// to run in the background and cannot contain any stack references.
+///
+/// # Panics
+///
+/// This function has the same panic behavior as `future_to_promise`.
+pub fn spawn_local<F>(future: F)
+where
+    F: Future<Output = ()> + 'static,
+{
+    struct Task {
+        future: Mutex<Option<Pin<Box<dyn Future<Output = ()> + 'static>>>>,
```

Comment:
Since the `Future` here is fundamentally non-`Send`, is the `Mutex`/`Atomic*` required or could they be replaced with non-threadsafe variants?

Reply:
Also, orthogonally, I wasn't quite sure why I initially saw `Option` here, but is the idea that you can drop the future at any time but there may still be lingering handles to the `Task` for awhile longer while events are processed throughout the system?

Reply:
The atomic stuff isn't needed, it's just a holdover from when I tried to make it thread safe (and then later gave up).

The reason for the `Option` is because the Future might return `Poll::Ready` and also call `waker.wake()`. That would end up causing an infinite loop, so the `Option` lets us avoid that.

Sure, that's technically against the Future contract, but I like to make the code rock solid so it can survive even bad Futures.

It also has the benefit that it can drop the Future immediately when it returns `Poll::Ready` (rather than at some potentially far-future time when the `Task` is eventually dropped).

Because the `Waker` contains an `Arc<Task>`, it's theoretically possible that the Future will *never* get dropped (if someplace somewhere keeps a hold of the `Waker`). So the `Option` lets us fix that.